crates/server/src/server/server_future.rs

```diff
@@ -553,54 +561,172 @@ impl<T: RequestHandler> ServerFuture<T> {
+            is_dnssec = is_dnssec,
+            op = qop_code,
+            query = query_name,
+            qtype = query_type,
+            class = query_class,
+            qflags = qflags,
+        );
+
+        // The reporter will handle making sure to log the result of the request
+        let reporter = ReportingResponseHandler {
+            request_header: *request.header(),
+            query,
+            protocol,
+            src_addr,
+            handler: response_handler,
+        };
+
+        request_handler
+            .lock()
+            .await
+            .handle_request(request, reporter)
+            .await;
     };
 
-    info!(
-        "request: {} type: {:?} op_code: {:?} dnssec: {} {}",
-        request.message.id(),
-        request.message.message_type(),
-        request.message.op_code(),
-        request.message.edns().map_or(false, Edns::dnssec_ok),
-        request
-            .message
-            .queries()
-            .first()
-            .map(|q| q.original().to_string())
-            .unwrap_or_else(|| "empty_queries".to_string()),
-    );
-
-    request_handler
-        .lock()
-        .await
-        .handle_request(request, response_handler)
-        .await
+    // Attempt to decode the message
+    match MessageRequest::read(&mut decoder) {
+        Ok(message) => {
+            inner_handle_request(message, response_handler).await;
+        }
+        Err(ProtoError { kind, .. }) if kind.as_form_error().is_some() => {
+            // We failed to parse the request due to some issue in the message, but the header is available, so we can respond
+            let (header, error) = kind
+                .into_form_error()
+                .expect("as form_error already confirmed this is a FormError");
+            let query = LowerQuery::query(Query::default());
+
+            // debug for more info on why the message parsing failed
+            debug!(
+                "request:{id} src:{proto}://{addr}#{port} type:{message_type} {op}:FormError:{error}",
+                id = header.id(),
+                proto = protocol,
+                addr = src_addr.ip(),
+                port = src_addr.port(),
+                message_type= header.message_type(),
+                op = header.op_code(),
+                error = error,
+            );
+
+            // The reporter will handle making sure to log the result of the request
+            let mut reporter = ReportingResponseHandler {
+                request_header: header,
+                query,
+                protocol,
+                src_addr,
+                handler: response_handler,
+            };
+
+            let response = MessageResponseBuilder::new(None);
+            let result = reporter
+                .send_response(response.error_msg(&header, ResponseCode::FormErr))
+                .await;
+
+            if let Err(e) = result {
+                warn!("failed to return FormError to client: {}", e);
```

Comment:
I wonder if we should make this a debug! as well as the parse failure below, to avoid overflowing the logs? at least downgrade to info!